File: src/models/PromoCode.js
const mongoose = require('mongoose');
const { hashCode, extractPrefix } = require('../utils/promoCode.util');

const promoCodeSchema = new mongoose.Schema({
  // Hash du code (jamais le code en clair)
  codeHash: {
    type: String,
    required: [true, 'Le hash du code est requis'],
    unique: true,
    index: true,
    trim: true
  },
  // Préfixe pour faciliter la recherche (ex: "A3B2" pour RT-A3B2-XXXX-XXXX)
  codePrefix: {
    type: String,
    required: true,
    trim: true,
    uppercase: true,
    maxlength: [10, 'Le préfixe ne peut pas dépasser 10 caractères'],
    index: true
  },
  // Type de code promotionnel
  type: {
    type: String,
    required: [true, 'Le type de code est requis'],
    enum: {
      values: ['DISCOUNT_PERCENT', 'GRANT_PREMIUM_MONTHS', 'GRANT_PREMIUM_PERMANENT'],
      message: 'Le type doit être DISCOUNT_PERCENT, GRANT_PREMIUM_MONTHS ou GRANT_PREMIUM_PERMANENT'
    },
    index: true
  },
  // Pourcentage de réduction (si type = DISCOUNT_PERCENT)
  discountPercent: {
    type: Number,
    min: [1, 'Le pourcentage de réduction doit être entre 1 et 100'],
    max: [100, 'Le pourcentage de réduction doit être entre 1 et 100'],
    validate: {
      validator: function(value) {
        if (this.type === 'DISCOUNT_PERCENT') {
          return value != null && value >= 1 && value <= 100;
        }
        return value == null;
      },
      message: 'discountPercent est requis et doit être entre 1 et 100 pour DISCOUNT_PERCENT'
    }
  },
  // Nombre de mois Premium (si type = GRANT_PREMIUM_MONTHS)
  premiumMonths: {
    type: Number,
    min: [1, 'Le nombre de mois Premium doit être au moins 1'],
    validate: {
      validator: function(value) {
        if (this.type === 'GRANT_PREMIUM_MONTHS') {
          return value != null && value >= 1;
        }
        return value == null;
      },
      message: 'premiumMonths est requis et doit être >= 1 pour GRANT_PREMIUM_MONTHS'
    }
  },
  // Limite d'utilisation (nombre de fois que le code peut être utilisé)
  usageLimit: {
    type: Number,
    default: 1,
    min: [1, 'La limite d\'utilisation doit être au moins 1']
  },
  // Nombre de fois que le code a été utilisé
  usedCount: {
    type: Number,
    default: 0,
    min: [0, 'Le compteur d\'utilisation ne peut pas être négatif']
  },
  // Historique des utilisations (pour audit)
  usedBy: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    usedAt: {
      type: Date,
      default: Date.now
    },
    // Détails optionnels (ex: montant économisé, mois accordés, etc.)
    details: {
      type: mongoose.Schema.Types.Mixed,
      default: null
    }
  }],
  // Dates de validité
  validFrom: {
    type: Date,
    default: null
  },
  validUntil: {
    type: Date,
    default: null
  },
  // Statut actif/inactif
  isActive: {
    type: Boolean,
    default: true,
    index: true
  },
  // Créateur (admin)
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Le créateur est requis']
  },
  // Métadonnées libres (ex: campagne, description, etc.)
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  }
}, {
  timestamps: true // Ajoute createdAt et updatedAt automatiquement
});

// Index composé pour les requêtes admin (codes actifs et valides)
promoCodeSchema.index({ isActive: 1, validUntil: 1 });

// Index pour recherche par préfixe
promoCodeSchema.index({ codePrefix: 1, isActive: 1 });

// Méthode statique pour créer un code promotionnel
promoCodeSchema.statics.createFromPlainCode = async function(plainCode, data, createdBy) {
  const codeHash = hashCode(plainCode);
  const codePrefix = extractPrefix(plainCode);
  
  // Vérifier que le code n'existe pas déjà
  const existing = await this.findOne({ codeHash });
  if (existing) {
    throw new Error('Ce code promotionnel existe déjà');
  }
  
  const promoCode = new this({
    codeHash,
    codePrefix,
    type: data.type,
    discountPercent: data.discountPercent,
    premiumMonths: data.premiumMonths,
    usageLimit: data.usageLimit || 1,
    validFrom: data.validFrom,
    validUntil: data.validUntil,
    isActive: data.isActive !== undefined ? data.isActive : true,
    createdBy,
    metadata: data.metadata || {}
  });
  
  return await promoCode.save();
};

// Méthode pour vérifier si le code est valide
promoCodeSchema.methods.isValid = function() {
  if (!this.isActive) {
    return false;
  }
  
  if (this.usedCount >= this.usageLimit) {
    return false;
  }
  
  const now = new Date();
  
  if (this.validFrom && now < this.validFrom) {
    return false;
  }
  
  if (this.validUntil && now > this.validUntil) {
    return false;
  }
  
  return true;
};

// Méthode pour vérifier si un utilisateur a déjà utilisé ce code
promoCodeSchema.methods.hasBeenUsedBy = function(userId) {
  if (!userId) {
    return false;
  }
  
  const userIdStr = userId.toString();
  return this.usedBy.some(usage => usage.userId.toString() === userIdStr);
};

// Méthode pour enregistrer l'utilisation du code
promoCodeSchema.methods.recordUsage = async function(userId, details = null) {
  if (!this.isValid()) {
    throw new Error('Ce code promotionnel n\'est plus valide');
  }
  
  if (this.hasBeenUsedBy(userId)) {
    throw new Error('Vous avez déjà utilisé ce code promotionnel');
  }
  
  this.usedBy.push({
    userId,
    usedAt: new Date(),
    details
  });
  
  this.usedCount += 1;
  
  return await this.save();
};

module.exports = mongoose.model('PromoCode', promoCodeSchema);

File: flutter_app/lib/models/admin/promo_code.dart
class AdminGeneratePromoCodesRequest {
  final String type; // 'DISCOUNT_PERCENT', 'GRANT_PREMIUM_MONTHS', 'GRANT_PREMIUM_PERMANENT'
  final int count;
  final int? discountPercent;
  final int? premiumMonths;
  final int? usageLimit;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final Map<String, dynamic>? metadata;

  AdminGeneratePromoCodesRequest({
    required this.type,
    required this.count,
    this.discountPercent,
    this.premiumMonths,
    this.usageLimit,
    this.validFrom,
    this.validUntil,
    this.metadata,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'type': type,
      'count': count,
    };

    if (discountPercent != null) {
      json['discountPercent'] = discountPercent;
    }
    if (premiumMonths != null) {
      json['premiumMonths'] = premiumMonths;
    }
    if (usageLimit != null) {
      json['usageLimit'] = usageLimit;
    }
    if (validFrom != null) {
      json['validFrom'] = validFrom!.toIso8601String();
    }
    if (validUntil != null) {
      json['validUntil'] = validUntil!.toIso8601String();
    }
    if (metadata != null) {
      json['metadata'] = metadata;
    }

    return json;
  }
}

class GeneratedPromoCode {
  final String id;
  final String code;
  final String prefix;

  GeneratedPromoCode({
    required this.id,
    required this.code,
    required this.prefix,
  });

  factory GeneratedPromoCode.fromJson(Map<String, dynamic> json) {
    return GeneratedPromoCode(
      id: json['id'] ?? json['_id'] ?? '',
      code: json['code'] ?? '',
      prefix: json['prefix'] ?? '',
    );
  }
}

class PromoCodeListItem {
  final String id;
  final String codePrefix;
  final String type; // 'DISCOUNT_PERCENT', 'GRANT_PREMIUM_MONTHS', 'GRANT_PREMIUM_PERMANENT'
  final bool isActive;
  final int usedCount;
  final int usageLimit;
  final DateTime? validUntil;
  final int? discountPercent;
  final int? premiumMonths;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  PromoCodeListItem({
    required this.id,
    required this.codePrefix,
    required this.type,
    required this.isActive,
    required this.usedCount,
    required this.usageLimit,
    this.validUntil,
    this.discountPercent,
    this.premiumMonths,
    this.createdAt,
    this.updatedAt,
  });

  factory PromoCodeListItem.fromJson(Map<String, dynamic> json) {
    return PromoCodeListItem(
      id: json['_id'] ?? json['id'] ?? '',
      codePrefix: json['codePrefix'] ?? '',
      type: json['type'] ?? '',
      isActive: json['isActive'] ?? true,
      usedCount: json['usedCount'] ?? 0,
      usageLimit: json['usageLimit'] ?? 1,
      validUntil: json['validUntil'] != null
          ? DateTime.parse(json['validUntil'])
          : null,
      discountPercent: json['discountPercent'],
      premiumMonths: json['premiumMonths'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  String get typeDisplayName {
    switch (type) {
      case 'DISCOUNT_PERCENT':
        return 'Réduction ${discountPercent ?? 0}%';
      case 'GRANT_PREMIUM_MONTHS':
        return 'Premium ${premiumMonths ?? 0} mois';
      case 'GRANT_PREMIUM_PERMANENT':
        return 'Premium illimité';
      default:
        return type;
    }
  }
}

class PromoCodesPage {
  final List<PromoCodeListItem> items;
  final int total;

  PromoCodesPage({
    required this.items,
    required this.total,
  });

  factory PromoCodesPage.fromJson(Map<String, dynamic> json) {
    final itemsList = (json['items'] as List<dynamic>?)
            ?.map((item) => PromoCodeListItem.fromJson(item))
            .toList() ??
        [];

    return PromoCodesPage(
      items: itemsList,
      total: json['total'] ?? 0,
    );
  }
}







